backend/lib/aspace_json_to_yale_container_mapper.rb

```diff
@@ -0,0 +1,175 @@
+class AspaceJsonToYaleContainerMapper
+
+  include JSONModel
+
+  def initialize(json)
+    @json = json
+  end
+
+
+  def call
+    @json['instances'].each do |instance|
+
+      if instance['sub_container']
+        # Just need to make sure there are no conflicting ArchivesSpace containers
+        instance.delete('container')
+        next
+      end
+
+      top_container = get_or_create_top_container(instance)
+
+      ensure_harmonious_values(TopContainer.to_jsonmodel(top_container), instance['container'])
+
+      instance['sub_container'] = {
+        'top_container' => {'ref' => top_container.uri},
+        'type_2' => instance['container']['type_2'],
+        'indicator_2' => instance['container']['indicator_2'],
+        'type_3' => instance['container']['type_3'],
+        'indicator_3' => instance['container']['indicator_3'],
+      }
+
+      # No need for the original value now.
+      instance.delete('container')
+    end
+  end
+
+
+  private
+
+
+  def get_or_create_top_container(instance)
+    container = instance['container']
+
+    result = (try_matching_barcode(container) ||
+              try_matching_indicator_within_series(container) ||
+              try_matching_indicator_within_resource(container))
+
+    if result
+      result
+    else
+      p "UNHANDLED MAPPING - generating new container"
+      p ({:instance => instance, :record => @json})
+      TopContainer.create_from_json(JSONModel(:top_container).from_hash('barcode' => SecureRandom.hex, 'indicator' => (rand() * 10000).to_i.to_s))
```

Comment:
Need to think about this... should probably be the default indicator (or the incoming indicator)